broker/api/broker.yaml

```
@@ -0,0 +1,978 @@
+                enum:
+                  - UNKNOWN
+                  - QUEUED
+                  - RUNNING
+                  - DELETING
+                  - SUCCEEDED
+                  - FAILED
+                example: QUEUED
+                description: Current condition/state of build. 
+              started_at:
+                type: string
+                format: date-time
+                example: 2021-06-11T17:32:28Z
+                description: Timestamp taken at the begining of build.
+              finished_at:
+                type: string
+                format: date-time
+                example: 2021-06-11T17:32:28Z
+                description: Timestamp taken when build is finished and ready to
+                 deploy.
+            required:
+              - status
+              - started_at
+              - finished_at
+    # BuildStates:
+    #   type: string
+    #   enum:
+    #     - UNKNOWN
+    #     - QUEUED
+    #     - RUNNING
+    #     - DELETING
+    #     - SUCCEEDED
+    #     - FAILED
+    BuildRegister:
+      type: object
+      description: Describes schema for registering build
+      properties:
+        images:
+          type: array
+          items: 
+            $ref: '#/components/schemas/Image'
+          description: List of image to be build. It can contain single or
+           multiple images depending on requirement of repository.
+        head_commit:
+          type: object
+          oneOf:
+            - $ref: '#/components/schemas/Branch'
+            - $ref: '#/components/schemas/Tag'
+          description: "Reference to the `commit` or `tag` which is to be used
+           for source code checkout"
+      required:
+          - images
+          - head_commit
+    Image:
+      type: object
+      description: Properties required for building docker image i.e. name/tag
+       and location of Dockerfile
+      properties:
+        name:
+          type: string
+          description: Image repository and tag used to build image.
+          example: akash7778/broker:0.0.1
+        location:
+          type: string
+          description: Location of Dockerfile relative to repository root
+           directory.
+          default: ./Dockerfile
+      required:
+            - name
+    Branch:
+      type: object    # Branch
+      description: 'Git branch: Used as reference for git checkout'
+      properties:
+            branch:
+                type: string
+                description: Name of branch to be used to checkout git
+                 repository.
+                example: development
+            commit_sha:
+                type: string
+                description: Git commit sha till where repository needs to be
+                 checkout by broker. Head commit will be used if `commit_sha`
+                  is not supplied in request.
+                default: <Head-commit>
+                example: 930fd5
+      required:
+        - branch
```

```suggestion
        - branch
        - commit_sha
```

Commit_sha should be made required

---

If `commit_sha` is not provided it will take top `commit` of the `branch` and checkout.

---

When I tried it, if `commit_sha` was not provided, the Broker was crashing.
I will test it again

---

I will also test it again

---

I just tried again. When I make the request without `commit_sha` I get a HTTP/500 error as a response and a `KeyError` in the logs:

```
[2021-08-23 05:25:03,673: ERROR   ] Traceback (most recent call last):\n  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1950, in full_dispatch_request\n    rv = self.dispatch_request()\n  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1936, in dispatch_request\n    return self.view_functions[rule.endpoint](**req.view_args)\n  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/decorator.py", line 48, in wrapper\n    response = function(request)\n  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/uri_parsing.py", line 144, in wrapper\n    response = function(request)\n  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/validation.py", line 184, in wrapper\n    response = function(request)\n  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/validation.py", line 384, in wrapper\n    return function(request)\n  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/response.py", line 103, in wrapper\n    response = function(request)\n  File "/usr/local/lib/python3.9/site-packages/connexion/decorators/parameter.py", line 121, in wrapper\n    return function(**kwargs)\n  File "/usr/local/lib/python3.9/site-packages/foca/utils/logging.py", line 61, in _wrapper\n    response = fn(*args, **kwargs)\n  File "/app/broker/ga4gh/broker/server.py", line 105, in postBuild\n    return register_builds(id, request.headers['X-Project-Access-Token'],\n  File "/app/broker/ga4gh/broker/endpoints/builds.py", line 105, in register_builds\n    commit=build_data['head_commit'][\nKeyError: 'commit_sha' [foca.errors.exceptions]
```

---

I'm working on it

---

Fixed this. Also added for only tag option.
But just now realized what if someone registers a build with the wrong `dockerfile` path... This is handled but while creating build... Right now I am opening an issue for this